db/core/file.c

```diff
@@ -194,14 +195,27 @@ tempesta_map_file(struct file *file, unsigned long len, int node)
 
 	get_file(file);
 
-	addr = kernel_read(file, (char *)ma->start, len, &off);
-	if (addr != len) {
-		TDB_ERR("Cannot read %lu bytes to addr %p, ret = %ld\n",
-			len, (void *)ma->start, addr);
-		fput(file);
-		__ma_free(ma);
-		addr = -EIO;
-		goto err;
+	while (copied < len) {
+		ssize_t bytes;
+		size_t to_read = len - off;
+
+		if (to_read > MAX_RW_COUNT)
+			to_read = MAX_RW_COUNT;
+
+		bytes = kernel_read(file, ((char *)ma->start) + copied, to_read,
+				    &off);
+		if (bytes < 0) {
+			addr = bytes;
+			TDB_ERR("Cannot read %li bytes to addr %p, ret = %ld\n",
+				to_read, (void *)ma->start + copied, addr);
+			fput(file);
+			__ma_free(ma);
+			goto err;
+		}
+
+		if (bytes == 0)
```

Comment:
It seems wrong. If we bytes == 0 before we read all (copied < len) we should return error